src/main/scala/org/broadinstitute/hail/driver/FilterAlleles.scala

```diff
@@ -0,0 +1,156 @@
+    val finalType = types.foldLeft(state.vds.vaSignature) { case (vas, (path, signature)) =>
+      val (newVas, i) = vas.insert(signature, path)
+      inserterBuilder += i
+      newVas
+    }
+    val inserters = inserterBuilder.result()
+
+    val keep = options.keep
+
+    def filterAllelesInVariant(v: Variant, va: Annotation): Option[(Variant, IndexedSeq[Int], Array[Int])] = {
+      var allZero = true
+      var dead = 0
+      val oldToNew =
+        coalesce(v.altAlleles)(v.altAlleles.size+1, (_, altIndex) => altIndex + 1, 0) { (_,_,_,index) =>
+            conditionEC.setAll(v, va, index)
+            if (Filter.keepThis(conditionE(), keep)) {
+              allZero = false
+              index - dead
+            } else {
+              dead += 1
+              0
+            } }
+
+      if (allZero)
+        Option.empty
+      else {
+        val survivorIndexMap = oldToNew.iterator
+            .zipWithIndex
+            .filter { case (newIdx, oldIdx) => oldIdx == 0 || newIdx != 0 }
+            .map(_._2)
+            .toArray
+
+        val altAlleles = oldToNew.iterator
+            .zipWithIndex
+            .filter { case (newIdx, _) => newIdx != 0 }
+            .map { case (_, idx) => v.altAlleles(idx-1) }
+            .toArray
+
+        Option((v.copy(altAlleles = altAlleles), survivorIndexMap : IndexedSeq[Int], oldToNew))
+      }
+    }
+
+    def updateAnnotation(v: Variant, va: Annotation, indexMap: IndexedSeq[Int]): Annotation = {
+      annotationEC.setAll(v, va, indexMap)
+      generators.zip(inserters).foldLeft(va) { case (va, (fn, inserter)) => inserter(va, fn()) }
+    }
+
+    def updateGenotypes(gs: Iterable[Genotype], oldToNew: Array[Int], survivorCount: Int): Iterable[Genotype] = {
+      def downcodeGtPair(gt: GTPair): GTPair =
+        GTPair.fromNonNormalized(oldToNew(gt.j), oldToNew(gt.k))
+      def downcodeGt(gt: Int): Int =
+        Genotype.gtIndex(downcodeGtPair(Genotype.gtPair(gt)))
+
+      def downcodeAd(ad: Array[Int]): Array[Int] = {
+        coalesce(ad)(survivorCount, (_, alleleIndex) => oldToNew(alleleIndex), 0) { (oldDepth, depth, _, _) =>
+          oldDepth + depth
+        }
+      }
+
+      def downcodePx(px: Array[Int]): Array[Int] = {
+        coalesce(px)(survivorCount*(survivorCount+1)/2, (_, gt) => downcodeGt(gt), Int.MaxValue) { (oldNll, nll, _, _) =>
+          min(oldNll, nll)
+        }
+      }
+
+      def downcodeGenotype(g: Genotype): Genotype = {
+        val px = g.px.map(downcodePx)
+        val gq = px.map(Genotype.gqFromPL)
+        Genotype(g.gt.map(downcodeGt),
+          g.ad.map(downcodeAd),
+          g.dp,
+          gq,
+          px,
+          g.fakeRef,
+          g.isDosage
+        )
+      }
+
+      gs.map(downcodeGenotype)
+    }
+
+    def updateOrFilterRow(v: Variant, va: Annotation, gs: Iterable[Genotype]): Option[(Variant, (Annotation, Iterable[Genotype]))] =
+      for ((newV, survivingIndexMap, oldToNew) <- filterAllelesInVariant(v, va);
```

Comment:
I think we elsewhere use `Option.map ( ... )` preferentially to `for (x <- Option) yield ...`